sentinelhub/api/batch_base.py

```diff
@@ -0,0 +1,98 @@
+"""
+Module containing shared code of Batch Process API and Batch Statistical API
+"""
+from abc import ABCMeta
+from enum import Enum
+from typing import Generic, Iterable, Optional, Sequence, TypeVar, Union
+
+from ..constants import RequestType
+from ..type_utils import Json
+from .base import SentinelHubService
+
+BatchRequestType = TypeVar("BatchRequestType", bound="BaseBatchRequest")  # pylint: disable=invalid-name
+RequestSpec = Union[str, dict, BatchRequestType]
+
+
+class BatchRequestStatus(Enum):
+    """An enum class with all possible batch request statuses"""
+
+    CREATED = "CREATED"
+    ANALYSING = "ANALYSING"
+    ANALYSIS_DONE = "ANALYSIS_DONE"
+    PROCESSING = "PROCESSING"
+    DONE = "DONE"
+    FAILED = "FAILED"
+    PARTIAL = "PARTIAL"
+    CANCELED = "CANCELED"
+
+
+class BatchUserAction(Enum):
+    """An enum class with all possible batch user actions"""
+
+    START = "START"
+    ANALYSE = "ANALYSE"
+    NONE = "NONE"
+    CANCEL = "CANCEL"
+
+
+class BaseBatchClient(SentinelHubService, Generic[BatchRequestType], metaclass=ABCMeta):
+    """Class containing common methods and helper functions for Batch Client classes"""
+
+    def _call_job(self, batch_request: RequestSpec, endpoint_name: str) -> Json:
+        """Makes a POST request to the service that triggers a processing job"""
+        request_id = self._parse_request_id(batch_request)
+        job_url = f"{self._get_job_endpoint_url(request_id)}/{endpoint_name}"
+        return self.client.get_json(url=job_url, request_type=RequestType.POST, use_session=True)
+
+    def _get_job_endpoint_url(self, request_id: Optional[str] = None) -> str:
+        """Creates a URL for the batch statistical endpoint"""
+        url = f"{self.service_url}"
+        if request_id:
+            return f"{url}/{request_id}"
+        return url
+
+    @staticmethod
+    def _parse_request_id(data: RequestSpec) -> str:
+        """Parses batch request id from multiple possible inputs"""
+        if isinstance(data, BaseBatchRequest):
+            return data.request_id
+        if isinstance(data, dict):
+            return data["id"]
+        if isinstance(data, str):
+            return data
+        raise ValueError(f"Expected a BatchRequest, dictionary or a string, got {data}.")
+
+
+class BaseBatchRequest:
```

Comment:
because dataclasses + inheritance + default values dont mix well. If the parent dataclass has an attribute with a default value then the child classes can only extend with default-valued attributes.

I could wrap this in a dataclass and just leave `error` without a default and then override the field with a new field + default in the child classes.

Reply:
Ah, you are right. This is an annoying thing with dataclasses.

But even now you don't assign `error` attribute a default value. So in the current implementation it might work?

Reply:
sadly this couldn't be done (easily). I tried many options but dataclass_json seems very inflexible when it comes to inheritance